### DISCUSSION TITLE
typical pipeline with parserino

### DISCUSSION THREAD

**Discussion by aferust** [trikko/serverino#17] [Q&A] [answered]
So, based on this example https://github.com/trikko/parserino/tree/master/examples/05_serve_html, we create our HTML pages with any editor, set some placeholders, then load and edit innerText dynamically. There should be capabilities of parserino and/or lexbor to pass js stuff. I should examine those capabilities, such as forms, etc.

**Comment by trikko** [comment, marked as answer]
For passing data from server to client you can use a hidden node with a custom data- attribute. Something like:

```html
...
<div id="mydata" data-content="">
...
```

```d
auto json = ....;
auto node = doc.byId("mydata");
node.setAttribute("data-content", json.toString);
```

For passing data from client to server you can use a webservice, of course. 
```js
fetch("https://yourserver/dothis")
.then((response) => response.json())
.then((json) => console.log(json));
```

**Reply by aferust** [reply]
It is challenging to guess the usage of parserino; at least my heuristic approach fails here in updating a table. The docs do not help at all.
```
...
<body>
    <h1>pholder</h1>
    <h2 id="sub1">pholder</h2>
    <table id="tbl1" class="ui celled table">
    </table>
</body>
...
```
```d
        auto table = doc.byId("tbl1"); // I expect Element to use ref semantics here. if I modify the table it should be updated automagically?
	table.appendChild(
		asFragment("<thead>
            <tr>
                <th>Material</th>
                <th>Quantity</th>
                <th>Unit price</th>
            </tr>
        </thead>")
	);
	auto tbody = doc.createElement("tbody");
	

	string[string][] tableData = [
		["Material": "Acrylic (Transparent)", "Quantity": "25", "Unit price": "$2.90"],
		["Material": "Plywood (Birch)", "Quantity": "50", "Unit price": "$1.25"],
		["Material": "Laminate (Gold on Blue)", "Quantity": "10", "Unit price": "$2.35"]
	];
	foreach (row; tableData)
	{
		tbody.appendChild(
			asFragment(format!"<tr><td>%s</td><td>%s</td><td>%s</td></tr>"(row["Material"], row["Quantity"], row["Unit price"]))
		);
	}
	table.appendChild(tbody);
	doc.body.appendChild(table); // this is stupid, but I tried
	
	output ~= doc; 
```

Where is my table? Only

```
<html>
<body>
<!--StartFragment-->
<table id="tbl1" class="ui celled table">
--
  | <tbody></tbody></table></body></html>

<!--EndFragment-->
</body>
</html>
```

**Reply by aferust** [reply]
if i remove asFragment, this time it prints escaped string of table contents

**Reply by aferust** [reply]
ok I figured this out:

```d
	auto table = doc.byId("tbl1");
	auto thead = doc.createElement("thead");
	thead.innerHTML =
		asFragment("<thead>
            <tr>
                <th>Material</th>
                <th>Quantity</th>
                <th>Unit price</th>
            </tr>
        </thead>");
	
	auto tbody = doc.createElement("tbody");

	string[string][] tableData = [
		["Material": "Acrylic (Transparent)", "Quantity": "25", "Unit price": "$2.90"],
		["Material": "Plywood (Birch)", "Quantity": "50", "Unit price": "$1.25"],
		["Material": "Laminate (Gold on Blue)", "Quantity": "10", "Unit price": "$2.35"]
	];
	string inner;
	foreach (row; tableData)
	{
		inner ~= asFragment(format!"<tr><td>%s</td><td>%s</td><td>%s</td></tr>"(row["Material"], row["Quantity"], row["Unit price"]));
	}
	tbody.innerHTML = inner;
	
	table.appendChild(thead);
	table.appendChild(tbody);

	doc.writeln;
	output ~= doc;
```

**Reply by trikko** [reply]
I use another approach:
- I create the table in html with a line inside
- I clone the line and fill it with query (for each line I need!)
- I append the new cloned line to the table
- I remove the cloned line

If I don't need the table, I simply remove it.

Anyway it's a bit offtopic :)

**Reply by trikko** [reply]
(I create a full page, with all placeholders, tables, etc. and then I remove what I do not need)

**Reply by trikko** [reply]
Check this example where I use a full template, removing a part I didn't need ("pricing" session)
https://github.com/trikko/parserino/tree/master/examples/05_serve_html

**Reply by aferust** [reply]
Your approach must be very suitable for production. I was just experimenting to learn how things work. 

It would be better if innerHtml were to beautify my uneven whitespaces.

**Reply by trikko** [reply]
Long story short about uneven whitespaces: this is not done because actually you're going to modify the html tree (doc_with_space != doc_without_space also if it is rendered in the same way).

**Reply by aferust** [reply]
On the other hand, one can also remove all white spaces to minify a bit. Not a big deal after all.

**Reply by trikko** [reply]
Yes it's not that complex. But probably is not even that useful, considering that most web servers like nginx support gzip compression (and also minifying, I guess)
